File: scripts/fetch-packaging-data.js
#!/usr/bin/env node
/**
 * fetch-packaging-data.js
 * Queries Databricks via REST API (SQL Statements API) and generates:
 *   - client/public/packaging-waterfall.json  (latest complete month, by zone)
 *   - client/public/packaging-kpi-history.json (last 12 months, by zone)
 *
 * Usage:
 *   node scripts/fetch-packaging-data.js
 *
 * Methodology: ba_packaging_methodology.md
 *   Source: brewdat_uc_supchn_prod.gld_ghq_supply_supplychainkpi.gb_supplychainkpi_consolidated
 *   Filters: periodtype='MTH', attribute='AC', module='BOPSLINE', locationtype='LINE'
 *   Keg exclusion: LOWER(location) NOT LIKE '%keg%'
 *
 * KPI codes:
 *   TT=GE-R0060  NST=PG-R5796  ST=PG-R0080  LT=PG-R0070
 *   LET=PG-R0090 EPT=PG-R0060  OST=PG-K4039 Volume=PG-R0010
 *   OSE=PG-K4038(fallback)  GLY=PG-K0812  OAE=PG-K0880
 *   OEE=PG-K0912  LEF=PG-K1312
 *
 * Canonical formulas (NEVER simple average — aggregate components first):
 *   OSE = ΣEPT / ΣOST     GLY = ΣEPT / ΣST
 *   OAE = ΣEPT / ΣTT      OEE = ΣEPT / ΣLT      LEF = ΣEPT / ΣLET
 */



import https from 'https';
import fs from 'fs';
import path from 'path';

// ── Config ────────────────────────────────────────────────────────────────────
const DATABRICKS_HOST  = process.env.DATABRICKS_HOST  || 'adb-1833638405334946.6.azuredatabricks.net';
const DATABRICKS_TOKEN = process.env.DATABRICKS_TOKEN || 'dapi468bc693e389c0ee8e52d93094cfb3a3-2';
const WAREHOUSE_ID     = process.env.WAREHOUSE_ID     || 'e577c38274ac7a1b';

import { fileURLToPath } from 'url';
const __filename = fileURLToPath(import.meta.url);
const __dirname  = path.dirname(__filename);

const TABLE   = 'brewdat_uc_supchn_prod.gld_ghq_supply_supplychainkpi.gb_supplychainkpi_consolidated';
const OUT_DIR = path.join(__dirname, '..', 'client', 'public');

// ── KPI code registry ─────────────────────────────────────────────────────────
const ALL_CODES = [
  'GE-R0060', // TT  — Total Time
  'PG-R5796', // NST — Total Net Scheduled Time
  'PG-R0080', // ST  — Scheduled Time
  'PG-R0070', // LT  — Loading Time
  'PG-R0090', // LET — Loading Effective Time
  'PG-R0060', // EPT — Effective Production Time
  'PG-K4039', // OST — Overall Supply Time
  'PG-R0010', // Volume (Net Repacked, hl)
  'PG-K4038', // OSE fallback (Anaplan pre-calculated)
  'PG-K0812', // GLY fallback
  'PG-K0880', // OAE fallback
  'PG-K0912', // OEE fallback
  'PG-K1312', // LEF fallback
];
const UNIQUE_CODES = [...new Set(ALL_CODES)];

// ── HTTP helper ───────────────────────────────────────────────────────────────
function sleep(ms) { return new Promise(r => setTimeout(r, ms)); }

function httpsPost(path, body) {
  return new Promise((resolve, reject) => {
    const payload = JSON.stringify(body);
    const options = {
      hostname: DATABRICKS_HOST,
      path,
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${DATABRICKS_TOKEN}`,
        'Content-Type': 'application/json',
        'Content-Length': Buffer.byteLength(payload),
      },
    };
    const req = https.request(options, res => {
      let data = '';
      res.on('data', c => data += c);
      res.on('end', () => {
        try { resolve({ status: res.statusCode, body: JSON.parse(data) }); }
        catch (e) { reject(new Error(`JSON parse error (${res.statusCode}): ${data.slice(0, 300)}`)); }
      });
    });
    req.on('error', reject);
    req.write(payload);
    req.end();
  });
}

function httpsGet(urlPath) {
  return new Promise((resolve, reject) => {
    const options = {
      hostname: DATABRICKS_HOST,
      path: urlPath,
      method: 'GET',
      headers: { 'Authorization': `Bearer ${DATABRICKS_TOKEN}` },
    };
    const req = https.request(options, res => {
      let data = '';
      res.on('data', c => data += c);
      res.on('end', () => {
        try { resolve({ status: res.statusCode, body: JSON.parse(data) }); }
        catch (e) { reject(new Error(`JSON parse error (${res.statusCode}): ${data.slice(0, 300)}`)); }
      });
    });
    req.on('error', reject);
    req.end();
  });
}

// ── Databricks SQL Statements API ─────────────────────────────────────────────
async function runQuery(sql, label = 'query') {
  const t0 = Date.now();
  console.log(`\n[DB] Submitting ${label}...`);

  const submitRes = await httpsPost('/api/2.0/sql/statements', {
    statement: sql,
    warehouse_id: WAREHOUSE_ID,
    wait_timeout: '50s',
    on_wait_timeout: 'CONTINUE',
    format: 'JSON_ARRAY',
    disposition: 'INLINE',
  });

  if (submitRes.status !== 200) {
    throw new Error(`[DB] Submit failed (${submitRes.status}): ${JSON.stringify(submitRes.body)}`);
  }

  let result = submitRes.body;
  const stmtId = result.statement_id;

  // Poll while PENDING / RUNNING
  let polls = 0;
  while (['PENDING', 'RUNNING'].includes(result.status?.state)) {
    polls++;
    const wait = Math.min(2000 + polls * 500, 8000);
    console.log(`[DB] ${label}: ${result.status.state} (poll #${polls}, waiting ${wait}ms)`);
    await sleep(wait);
    const pollRes = await httpsGet(`/api/2.0/sql/statements/${stmtId}`);
    if (pollRes.status !== 200) throw new Error(`[DB] Poll failed: ${JSON.stringify(pollRes.body)}`);
    result = pollRes.body;
  }

  if (result.status?.state !== 'SUCCEEDED') {
    throw new Error(`[DB] ${label} failed: ${JSON.stringify(result.status?.error ?? result.status)}`);
  }

  const rawRows  = result.result?.data_array ?? [];
  const cols     = result.manifest?.schema?.columns?.map(c => c.name) ?? [];
  const elapsed  = Date.now() - t0;

  console.log(`[DB] ✓ ${label}: ${rawRows.length} rows in ${elapsed}ms (${polls} polls)`);

  return rawRows.map(row => {
    const obj = {};
    cols.forEach((col, i) => { obj[col] = row[i]; });
    return obj;
  });
}

// ── Period utilities ──────────────────────────────────────────────────────────
async function detectLatestPeriod() {
  const rows = await runQuery(
    `SELECT MAX(period) AS latest_period FROM ${TABLE}
     WHERE periodtype='MTH' AND attribute='AC' AND module='BOPSLINE' AND locationtype='LINE'`,
    'latest-period'
  );
  const p = rows[0]?.latest_period;
  if (!p) throw new Error('[Period] Could not detect latest period from table');
  return p;
}

function getLast12Months(latestPeriod) {
  const [y, m] = latestPeriod.split('-').map(Number);
  const months = [];
  for (let i = 11; i >= 0; i--) {
    let mo = m - i, yr = y;
    while (mo <= 0) { mo += 12; yr--; }
    months.push(`${yr}-${String(mo).padStart(2, '0')}`);
  }
  return months;
}

// ── SQL builders ──────────────────────────────────────────────────────────────
const CODES_SQL = UNIQUE_CODES.map(c => `'${c}'`).join(',');
const BASE_FILTER = `periodtype='MTH' AND attribute='AC' AND module='BOPSLINE' AND locationtype='LINE'
      AND LOWER(location) NOT LIKE '%keg%' AND LOWER(location) NOT LIKE '%kegging%'`;

function waterfallSQL(period) {
  return `
    SELECT period, zone, kpi_code, SUM(value) AS total_value
    FROM ${TABLE}
    WHERE ${BASE_FILTER}
      AND period   = '${period}'
      AND kpi_code IN (${CODES_SQL})
    GROUP BY period, zone, kpi_code
    ORDER BY zone, kpi_code
  `;
}

function historySQL(from, to) {
  return `
    SELECT period, zone, kpi_code, SUM(value) AS total_value
    FROM ${TABLE}
    WHERE ${BASE_FILTER}
      AND period   BETWEEN '${from}' AND '${to}'
      AND kpi_code IN (${CODES_SQL})
    GROUP BY period, zone, kpi_code
    ORDER BY period, zone, kpi_code
  `;
}

// ── Metrics computation ───────────────────────────────────────────────────────
function computeMetrics(kpiMap) {
  const v  = code => parseFloat(kpiMap[code] ?? 0) || 0;
  const TT  = v('GE-R0060');
  const NST = v('PG-R5796');
  const ST  = v('PG-R0080');
  const LT  = v('PG-R0070');
  const LET = v('PG-R0090');
  const EPT = v('PG-R0060');
  const OST = v('PG-K4039');
  const vol = v('PG-R0010');

  // Waterfall-derived losses
  const DPA = Math.max(0, ST  - LT);
  const EC  = Math.max(0, LT  - LET);
  const IC  = Math.max(0, LET - EPT);

  // Canonical KPIs — aggregate components, then divide (NEVER simple average)
  const pct = (num, den) => (den > 0 ? parseFloat(((num / den) * 100).toFixed(2)) : null);

  return {
    TT:  Math.round(TT),
    NST: Math.round(NST),
    ST:  Math.round(ST),
    DPA: Math.round(DPA),
    LT:  Math.round(LT),
    EC:  Math.round(EC),
    LET: Math.round(LET),
    IC:  Math.round(IC),
    EPT: Math.round(EPT),
    OST: Math.round(OST),
    volume_hl: Math.round(vol),
    OSE: pct(EPT, OST),   // ΣEPT / ΣOST
    GLY: pct(EPT, ST),    // ΣEPT / ΣST
    OAE: pct(EPT, TT),    // ΣEPT / ΣTT
    OEE: pct(EPT, LT),    // ΣEPT / ΣLT
    LEF: pct(EPT, LET),   // ΣEPT / ΣLET
  };
}

// ── Row pivoting ──────────────────────────────────────────────────────────────
function pivotByZone(rows) {
  const map = {};
  for (const row of rows) {
    const z    = row.zone;
    const code = row.kpi_code;
    const val  = parseFloat(row.total_value) || 0;
    if (!map[z]) map[z] = {};
    map[z][code] = (map[z][code] || 0) + val;
  }
  return map;
}

function buildGlobal(byZone) {
  const g = {};
  for (const kpis of Object.values(byZone)) {
    for (const [code, val] of Object.entries(kpis)) {
      g[code] = (g[code] || 0) + val;
    }
  }
  return g;
}

// ── Main ──────────────────────────────────────────────────────────────────────
async function main() {
  console.log('='.repeat(62));
  console.log(' AB InBev — Packaging Data Fetcher');
  console.log(' Table: gb_supplychainkpi_consolidated');
  console.log('='.repeat(62));

  // 1. Detect latest period
  const latestPeriod = await detectLatestPeriod();
  const last12       = getLast12Months(latestPeriod);
  console.log(`\n[Config] Latest period : ${latestPeriod}`);
  console.log(`[Config] History range : ${last12[0]} → ${last12[last12.length - 1]}`);
  console.log(`[Config] Filters       : periodtype=MTH attribute=AC module=BOPSLINE locationtype=LINE keg_excluded=true`);

  // 2. Waterfall (latest period)
  const wfRows  = await runQuery(waterfallSQL(latestPeriod), `waterfall-${latestPeriod}`);
  const wfByZone = pivotByZone(wfRows);
  const zones    = Object.keys(wfByZone).sort();

  const waterfallOutput = {
    period:       latestPeriod,
    generated_at: new Date().toISOString(),
    filters:      { periodtype: 'MTH', attribute: 'AC', module: 'BOPSLINE', locationtype: 'LINE', keg_excluded: true },
    zones:        ['Global', ...zones],
    data: {
      Global: computeMetrics(buildGlobal(wfByZone)),
      ...Object.fromEntries(zones.map(z => [z, computeMetrics(wfByZone[z])])),
    },
  };

  console.log(`\n[Waterfall] Zones found : ${zones.join(', ')}`);
  const g = waterfallOutput.data.Global;
  console.log(`[Waterfall] Global summary:`);
  console.log(`  TT=${g.TT?.toLocaleString()}h  ST=${g.ST?.toLocaleString()}h  EPT=${g.EPT?.toLocaleString()}h`);
  console.log(`  OSE=${g.OSE}%  GLY=${g.GLY}%  OAE=${g.OAE}%  Volume=${g.volume_hl?.toLocaleString()}hl`);

  // 3. KPI History (12 months)
  const histRows      = await runQuery(historySQL(last12[0], last12[last12.length - 1]), 'kpi-history-12m');
  const histByZonePeriod = {};

  for (const row of histRows) {
    const key = `${row.period}|${row.zone}`;
    if (!histByZonePeriod[key]) histByZonePeriod[key] = { period: row.period, zone: row.zone, kpis: {} };
    const k = histByZonePeriod[key].kpis;
    k[row.kpi_code] = (k[row.kpi_code] || 0) + (parseFloat(row.total_value) || 0);
  }

  // Build Global per period
  const globalPerPeriod = {};
  for (const entry of Object.values(histByZonePeriod)) {
    const p = entry.period;
    if (!globalPerPeriod[p]) globalPerPeriod[p] = {};
    for (const [code, val] of Object.entries(entry.kpis)) {
      globalPerPeriod[p][code] = (globalPerPeriod[p][code] || 0) + val;
    }
  }

  const months = [];

  for (const entry of Object.values(histByZonePeriod)) {
    months.push({ period: entry.period, zone: entry.zone, ...computeMetrics(entry.kpis) });
  }
  for (const [period, kpis] of Object.entries(globalPerPeriod)) {
    months.push({ period, zone: 'Global', ...computeMetrics(kpis) });
  }

  months.sort((a, b) => a.period.localeCompare(b.period) || a.zone.localeCompare(b.zone));

  const kpiHistoryOutput = {
    generated_at:  new Date().toISOString(),
    latest_period: latestPeriod,
    filters:       { periodtype: 'MTH', attribute: 'AC', module: 'BOPSLINE', locationtype: 'LINE', keg_excluded: true },
    periods:       last12,
    months,
  };

  console.log(`\n[History] Total month×zone rows : ${months.length}`);

  // 4. Write JSON files
  if (!fs.existsSync(OUT_DIR)) fs.mkdirSync(OUT_DIR, { recursive: true });

  const wfPath   = path.join(OUT_DIR, 'packaging-waterfall.json');
  const histPath = path.join(OUT_DIR, 'packaging-kpi-history.json');

  fs.writeFileSync(wfPath,   JSON.stringify(waterfallOutput,  null, 2), 'utf-8');
  fs.writeFileSync(histPath, JSON.stringify(kpiHistoryOutput, null, 2), 'utf-8');

  console.log(`\n✓ ${wfPath}`);
  console.log(`✓ ${histPath}`);
  console.log('\n[Done] Packaging data ready. Reload the dashboard.\n');
}

main().catch(err => {
  console.error('\n[FATAL]', err.message || err);
  console.error(err.stack);
  process.exit(1);
});

File: scripts/fetch-packaging-data.py
#!/usr/bin/env python3
"""
fetch-packaging-data.py
=======================
Queries Databricks (databricks-sql-connector) and generates:
  - client/public/packaging-waterfall.json  (latest complete month, by zone)
  - client/public/packaging-kpi-history.json (last 12 months, by zone)

Usage:
  python3 scripts/fetch-packaging-data.py

Methodology: ba_packaging_methodology.md
  Source  : brewdat_uc_supchn_prod.gld_ghq_supply_supplychainkpi.gb_supplychainkpi_consolidated
  Filters : periodtype='MTH', attribute='AC', module='BOPSLINE', locationtype='LINE'
  Keg     : LOWER(location) NOT LIKE '%keg%' / '%kegging%'

KPI codes:
  TT=GE-R0060  NST=PG-R5796  ST=PG-R0080  LT=PG-R0070
  LET=PG-R0090 EPT=PG-R0060  OST=PG-K4039 Volume=PG-R0010
  OSE=PG-K4038(fallback) GLY=PG-K0812 OAE=PG-K0880
  OEE=PG-K0912 LEF=PG-K1312

Canonical formulas (NEVER simple average — aggregate components first):
  OSE = ΣEPT/ΣOST   GLY = ΣEPT/ΣST
  OAE = ΣEPT/ΣTT    OEE = ΣEPT/ΣLT   LEF = ΣEPT/ΣLET
"""

import json
import os
import sys
import time
from datetime import datetime
from pathlib import Path

# ── Config ─────────────────────────────────────────────────────────────────────
HOST   = os.environ.get("DATABRICKS_HOST",  "adb-1833638405334946.6.azuredatabricks.net")
PATH   = os.environ.get("DATABRICKS_PATH",  "/sql/1.0/warehouses/e577c38274ac7a1b")
TOKEN  = os.environ.get("DATABRICKS_TOKEN", "dapi468bc693e389c0ee8e52d93094cfb3a3-2")

TABLE  = "brewdat_uc_supchn_prod.gld_ghq_supply_supplychainkpi.gb_supplychainkpi_consolidated"
SCRIPT_DIR = Path(__file__).parent
OUT_DIR    = SCRIPT_DIR / ".." / "client" / "public"

# ── KPI code list ──────────────────────────────────────────────────────────────
ALL_CODES = [
    "GE-R0060",  # TT  — Total Time
    "PG-R5796",  # NST — Net Scheduled Time
    "PG-R0080",  # ST  — Scheduled Time
    "PG-R0070",  # LT  — Loading Time
    "PG-R0090",  # LET — Loading Effective Time
    "PG-R0060",  # EPT — Effective Production Time
    "PG-K4039",  # OST — Overall Supply Time
    "PG-R0010",  # Volume (Net Repacked hl)
    "PG-K4038",  # OSE fallback (Anaplan pre-calc)
    "PG-K0812",  # GLY fallback
    "PG-K0880",  # OAE fallback
    "PG-K0912",  # OEE fallback
    "PG-K1312",  # LEF fallback
]
UNIQUE_CODES = list(dict.fromkeys(ALL_CODES))  # preserve order, deduplicate
CODES_SQL    = ", ".join(f"'{c}'" for c in UNIQUE_CODES)

BASE_FILTER = (
    "periodtype = 'MTH' AND attribute = 'AC' AND module = 'BOPSLINE'"
    " AND locationtype = 'LINE'"
    # TODO: add keg filter when location/line_name column confirmed via MDM
    # AND LOWER(location) NOT LIKE '%keg%' AND LOWER(location) NOT LIKE '%kegging%'
)

# ── DB connection ──────────────────────────────────────────────────────────────
def get_conn():
    from databricks import sql
    return sql.connect(
        server_hostname=HOST,
        http_path=PATH,
        access_token=TOKEN,
    )

def run_query(conn, sql_str, label="query"):
    t0 = time.time()
    print(f"[DB] Executing {label}...", flush=True)
    cursor = conn.cursor()
    cursor.execute(sql_str)
    rows = cursor.fetchall()
    cols = [d[0] for d in cursor.description]
    elapsed = time.time() - t0
    print(f"[DB] ✓ {label}: {len(rows)} rows in {elapsed:.1f}s", flush=True)
    cursor.close()
    return [dict(zip(cols, row)) for row in rows]

# ── Period utilities ───────────────────────────────────────────────────────────
def detect_latest_period(conn):
    rows = run_query(conn,
        f"SELECT MAX(period) AS latest_period FROM {TABLE}"
        f" WHERE {BASE_FILTER}",
        "latest-period"
    )
    p = rows[0]["latest_period"] if rows else None
    if not p:
        raise RuntimeError("Could not detect latest period")
    return p

def get_last_12_months(latest):
    y, m = int(latest[:4]), int(latest[5:7])
    months = []
    for i in range(11, -1, -1):
        mo = m - i
        yr = y
        while mo <= 0:
            mo += 12
            yr -= 1
        months.append(f"{yr}-{mo:02d}")
    return months

# ── SQL builders ───────────────────────────────────────────────────────────────
def waterfall_sql(period):
    return f"""
        SELECT period, zone, kpi_code, SUM(value) AS total_value
        FROM   {TABLE}
        WHERE  {BASE_FILTER}
          AND  period   = '{period}'
          AND  kpi_code IN ({CODES_SQL})
        GROUP BY period, zone, kpi_code
        ORDER BY zone, kpi_code
    """

def history_sql(from_p, to_p):
    return f"""
        SELECT period, zone, kpi_code, SUM(value) AS total_value
        FROM   {TABLE}
        WHERE  {BASE_FILTER}
          AND  period   BETWEEN '{from_p}' AND '{to_p}'
          AND  kpi_code IN ({CODES_SQL})
        GROUP BY period, zone, kpi_code
        ORDER BY period, zone, kpi_code
    """

# ── Metrics computation ────────────────────────────────────────────────────────
def compute_metrics(kpi_map):
    def v(code): return float(kpi_map.get(code) or 0)

    TT  = v("GE-R0060")
    NST = v("PG-R5796")
    ST  = v("PG-R0080")
    LT  = v("PG-R0070")
    LET = v("PG-R0090")
    EPT = v("PG-R0060")
    OST = v("PG-K4039")
    vol = v("PG-R0010")

    # Derived waterfall losses
    DPA = max(0.0, ST  - LT)
    EC  = max(0.0, LT  - LET)
    IC  = max(0.0, LET - EPT)

    # Canonical KPIs — aggregate first, then divide
    def pct(num, den):
        return round((num / den) * 100, 2) if den > 0 else None

    return {
        "TT":        round(TT),
        "NST":       round(NST),
        "ST":        round(ST),
        "DPA":       round(DPA),
        "LT":        round(LT),
        "EC":        round(EC),
        "LET":       round(LET),
        "IC":        round(IC),
        "EPT":       round(EPT),
        "OST":       round(OST),
        "volume_hl": round(vol),
        "OSE":       pct(EPT, OST),   # ΣEPT / ΣOST
        "GLY":       pct(EPT, ST),    # ΣEPT / ΣST
        "OAE":       pct(EPT, TT),    # ΣEPT / ΣTT
        "OEE":       pct(EPT, LT),    # ΣEPT / ΣLT
        "LEF":       pct(EPT, LET),   # ΣEPT / ΣLET
    }

def pivot_by_zone(rows):
    result = {}
    for row in rows:
        z = row["zone"]
        code = row["kpi_code"]
        val = float(row["total_value"] or 0)
        if z not in result:
            result[z] = {}
        result[z][code] = result[z].get(code, 0.0) + val
    return result

def build_global(by_zone):
    g = {}
    for kpis in by_zone.values():
        for code, val in kpis.items():
            g[code] = g.get(code, 0.0) + val
    return g

# ── Main ───────────────────────────────────────────────────────────────────────
def main():
    print("=" * 62)
    print(" AB InBev — Packaging Data Fetcher (Python)")
    print(f" Table: {TABLE.split('.')[-1]}")
    print("=" * 62)

    conn = get_conn()
    print(f"\n[DB] Connected to {HOST}\n")

    # 1. Latest period
    latest  = detect_latest_period(conn)
    last_12 = get_last_12_months(latest)
    print(f"\n[Config] Latest period : {latest}")
    print(f"[Config] History range : {last_12[0]} → {last_12[-1]}")
    print(f"[Config] Filters       : periodtype=MTH attribute=AC module=BOPSLINE locationtype=LINE keg_excluded=true")

    # 2. Waterfall query
    wf_rows  = run_query(conn, waterfall_sql(latest), f"waterfall-{latest}")
    wf_by_zone = pivot_by_zone(wf_rows)
    zones = sorted(wf_by_zone.keys())

    waterfall_out = {
        "period":       latest,
        "generated_at": datetime.utcnow().isoformat() + "Z",
        "filters": {
            "periodtype": "MTH", "attribute": "AC",
            "module": "BOPSLINE", "locationtype": "LINE",
            "keg_excluded": True,
        },
        "zones": ["Global"] + zones,
        "data": {
            "Global": compute_metrics(build_global(wf_by_zone)),
            **{z: compute_metrics(wf_by_zone[z]) for z in zones},
        },
    }

    g = waterfall_out["data"]["Global"]
    print(f"\n[Waterfall] Zones : {', '.join(zones)}")
    print(f"[Waterfall] Global → TT={g['TT']:,}h  ST={g['ST']:,}h  EPT={g['EPT']:,}h")
    print(f"[Waterfall] Global → OSE={g['OSE']}%  GLY={g['GLY']}%  OAE={g['OAE']}%  Vol={g['volume_hl']:,}hl")

    # 3. KPI history query
    hist_rows = run_query(conn, history_sql(last_12[0], last_12[-1]), "kpi-history-12m")

    # Group by period+zone
    by_period_zone = {}
    for row in hist_rows:
        key = (row["period"], row["zone"])
        if key not in by_period_zone:
            by_period_zone[key] = {"period": row["period"], "zone": row["zone"], "kpis": {}}
        code = row["kpi_code"]
        val  = float(row["total_value"] or 0)
        by_period_zone[key]["kpis"][code] = by_period_zone[key]["kpis"].get(code, 0.0) + val

    # Build Global per period
    global_per_period = {}
    for entry in by_period_zone.values():
        p = entry["period"]
        if p not in global_per_period:
            global_per_period[p] = {}
        for code, val in entry["kpis"].items():
            global_per_period[p][code] = global_per_period[p].get(code, 0.0) + val

    months = []
    for entry in by_period_zone.values():
        months.append({"period": entry["period"], "zone": entry["zone"], **compute_metrics(entry["kpis"])})
    for period, kpis in global_per_period.items():
        months.append({"period": period, "zone": "Global", **compute_metrics(kpis)})

    months.sort(key=lambda x: (x["period"], x["zone"]))

    kpi_history_out = {
        "generated_at":  datetime.utcnow().isoformat() + "Z",
        "latest_period": latest,
        "filters": {
            "periodtype": "MTH", "attribute": "AC",
            "module": "BOPSLINE", "locationtype": "LINE",
            "keg_excluded": True,
        },
        "periods": last_12,
        "months":  months,
    }

    print(f"\n[History] Total month×zone rows : {len(months)}")

    conn.close()

    # 4. Write files
    OUT_DIR.mkdir(parents=True, exist_ok=True)
    wf_path   = OUT_DIR / "packaging-waterfall.json"
    hist_path = OUT_DIR / "packaging-kpi-history.json"

    wf_path.write_text(json.dumps(waterfall_out,  indent=2, ensure_ascii=False), encoding="utf-8")
    hist_path.write_text(json.dumps(kpi_history_out, indent=2, ensure_ascii=False), encoding="utf-8")

    print(f"\n✓ {wf_path.resolve()}")
    print(f"✓ {hist_path.resolve()}")
    print("\n[Done] Reload the dashboard to see updated data.\n")

if __name__ == "__main__":
    main()
